Select components to add

sebastian-berlin-wmse opened the pull request:
Allows for selecting what to add for each project and general pages. The
flag `--components` shows a menu of project components. Only the
selected ones will be added for each processed project. The flag
`--prompt-add-pages` prompts before adding each general page, allowing
to skip pages.

Goals file is now only required if goal information is needed for the
selected pages.

Also:

* Added Tox to requirements.txt.
* Exclude Pywikibot files from Tox.

Bug: T297895

sebastian-berlin-wmse committed:
Select components to add

Allows for selecting what to add for each project and year pages. The
flag `--components` shows a menu of project components. Only the
selected ones will be added for each processed project. The flag
`--prompt-year-pages` prompts before adding each year page, allowing
to skip pages.

Goals file is now only required if goal information is needed for the
selected pages.

Bug: T297895

lokal-profil commented on const.py:
```diff
@@ -0,0 +1,8 @@
+from enum import Enum
```
`from enum import Enum, auto`

sebastian-berlin-wmse replied:
Separate lines is more legible (source: personal opinion🙂) and [recommended by PEP 8](https://www.python.org/dev/peps/pep-0008/#imports), though this case is mentioned as an acceptable exception.

lokal-profil replied:
What I've seen everywhere (and what is implemented in e.g. [isort](https://github.com/PyCQA/isort)) is
one import per line for different libraries
for "from X import" keep the same line if 3 or less, if more than three then wrap as a set and have one per line but indented.

lokal-profil commented on wiki.py:
```diff
@@ -308,6 +366,26 @@ def _make_year_title(self, raw_string):
+        if not self._prompt_year_pages:
+            return True
+
+        print('Add page "{}"? (y/N)'.format(self._make_year_title(title)))
```
why not use `input(...)` or `raw_input(...).lower()`?

sebastian-berlin-wmse replied:
Added prompt to input.

lokal-profil commented on const.py:
```diff
@@ -0,0 +1,8 @@
+
+
+class Components(Enum):
+    PROJECT_MAIN_PAGE = 1
```
any reason why this doesn't also use `auto()`?

sebastian-berlin-wmse replied:
Not sure. Maybe I thought it would star at zero and didn't want that, but it starts at one.

lokal-profil reviewed (commented):
First quick look. Not done with review

sebastian-berlin-wmse committed:
Review

lokal-profil commented on wiki.py:
```diff
@@ -27,90 +28,114 @@ class Wiki:
+        phab_name : str
+            Name of the project's Phabricator project.
+        """
+        if Components.PROJECT_MAIN_PAGE.value in self._components:
```
if self._components is None (because the argument wasn't passed) then this errors out

as does the `Components.CATEGORIES` test on L99

Either the tests all need to check something like `if self._components is None or Components.PROJECT_MAIN_PAGE.value in self._components` or the value of _components should be set to all allowed components when the argument is not passed.

lokal-profil commented on project_start.py:
```diff
@@ -38,6 +39,53 @@ def setup_logging(verbose):
+    for i, option in enumerate(options, start=1):
+        print("{}: {}".format(i, option))
+
+    selection_string = input()
```
Some text should be provided here. e.g. `input("Enter the numbers corresponding to the desired components, delimited by a single space character: ")` or if that is to long just `input("Enter the numbers delimited by a single space character: ")`

sebastian-berlin-wmse committed:
Review

sebastian-berlin-wmse commented:
I don't know why all the old code was added. The file changes are not what I have if I do a local diff against master. No idea why it's different here.

lokal-profil commented:
> I don't know why all the old code was added. The file changes are not what I have if I do a local diff against master. No idea why it's different here.

Quite shore you need to force push after a rebase (and you need to rebase to not get all the old commits )

sebastian-berlin-wmse commented:
Is there any way to revert it? Git told me do a `pull` before the `push` and that made a whole mess of it.

lokal-profil reviewed (changes requested):
Testing this I ran into a few error cases. 

`python project_start.py -d -o data/projektdata2022.tsv` with `components=1`
Results: in a crash at project_start.py:L287
Since the components does not require goals that crash should not have occurred. For the case of selecting a component which does need the goals the error message should reflect this.

`python project_start.py -d -o data/projektdata2022.tsv data/goals2021.tsv`  with `components=1`
Results: in a crash at wiki.py:L519

`python project_start.py -d -o -p "Organisationsutveckling 2022" data/projektdata2022.tsv data/goals2021.tsv` with `component=5`
Result: Fails to update page because it cannot find the project in the goals file. But component does not need goals info so should not be affected by this. (Leaving out the goals file results in the first crash listed above)

A missing feature (which can possibly be added later) is to make the general year pages (i.e. the ones like "Utbildning <YEAR>") a component. It would both make it possible to just run these or, more importantly, skip these when just running a single component update.

sebastian-berlin-wmse committed:
Review and fix some logging

sebastian-berlin-wmse commented:
> `python project_start.py -d -o data/projektdata2022.tsv data/goals2021.tsv` with `components=1` Results: in a crash at wiki.py:L519

That's not an error in the bot. It's because the page "Verksamhetsplan 2022/Tabell över program, strategi och mål" hasn't been created on the wiki yet. I added an exception saying this.

> `python project_start.py -d -o -p "Organisationsutveckling 2022" data/projektdata2022.tsv data/goals2021.tsv` with `component=5` Result: Fails to update page because it cannot find the project in the goals file. But component does not need goals info so should not be affected by this. (Leaving out the goals file results in the first crash listed above)

Added a note about this in the warning message. It can probably be solved in a way that works in these specific cases, but I don't think it's worth the time and effort.

> A missing feature (which can possibly be added later) is to make the general year pages (i.e. the ones like "Utbildning ") a component. It would both make it possible to just run these or, more importantly, skip these when just running a single component update.

You can do this with `--prompt-add-pages`, unless I'm missing something.

lokal-profil commented:
> You can do this with `--prompt-add-pages`, unless I'm missing something.

There are two big difference:
* If you don't want to add them you need to answer No the prompt for EACH page. With a component implementation I would expect the component to be for ALL of the general pages.
* There is no way of only updating the general pages (i.e. you cannot chose to have no components enabled to fall back on the general pages)

I still believe `--prompt-add-pages` is valuable even if a "general pages" component is added. But as I mentioned this can be left as a future feature.

lokal-profil commented on wiki.py:
```diff
@@ -512,6 +512,13 @@ def _get_programs(self):
+            title = operational_plan_page.title()
+            raise Exception(
+                f"Page '{title}' doesn't exist and is required to create "
+                "this page."
```
When the exception is raised it is not clear what "this page" references.

You probably need to let the exception bubble up one level and be re-raised where this context is known.

sebastian-berlin-wmse commented:
> There are two big difference:
> 
> * If you don't want to add them you need to answer No the prompt for EACH page. With a component implementation I would expect the component to be for ALL of the general pages.

Yes, but it's just tap enter ~15 times. What you're suggesting is a QoL update. I'm not against it, I just think that adding it may take more work than it's worth. Feel free to create a task for it and if it turns out that this is something that would save a lot of time we can look at it then.

> * There is no way of only updating the general pages (i.e. you cannot chose to have no components enabled to fall back on the general pages)

Yes, just enter nothing for the prompt and the general pages will still be processed.

sebastian-berlin-wmse committed:
Add messages with titles of pages that fail for page missing error
